doc/core/grpc-endpoint.md

```diff
@@ -0,0 +1,67 @@
+# gRPC Endpoints (`grpc_endpoint`) - TCP endpoint
+_Author: Sree Kuchibhotla (@sreecha) - Sep 2018_
+
+This document talks about `grpc_endpoint` abstraction and talks about the TCP endpoint's posix implementation in detail.
+
+### What is an endpoint ?
+`grpc_endpoint` is an abstraction used by gRPC to read/write bytes. There are different endpoint implementations depending on the underlying transport.
+
+
+### Endpoint interface
+
+The interface is documented well here: [`endpoint.h`](https://github.com/grpc/grpc/blob/v1.15.1/src/core/lib/iomgr/endpoint.h)
+
+### TCP endpoint
+
+TCP endpoint is implemented here: [tcp_posix.cc](https://github.com/grpc/grpc/blob/v1.15.1/src/core/lib/iomgr/tcp_posix.cc)
+
+#### Read
+
+``` C++
+// Implementation of grpc_endpoint_read() API
+void tcp_read(grpc_endpoint* ep, grpc_slice_buffer* slices, grpc_closure* cb) {
+ // 1. fd = Get grpc_fd from endpoint
+ // 2. Store read completion callback “cb” on the endpoint
+ // 3. Call polling engine API grpc_fd_notify_on_read(fd, tcp_handle_read)
+}
+```
+
+``` C++
+void tcp_handle_read() {
+  // 1. Do the actual reading by calling recvmsg()
```

Comment:
Again, if we already have read bytes, we might not need to call recvmsg again..

Reply:
Same response as above..I am intentionally keeping this simple..so that one can understand the overall (mostly-correct) flow :)